### ISSUE TITLE
rustls cannot reach a cluster through ip

### ISSUE THREAD

**Issue by clux** [kube-rs/kube#153]
```sh
kube api: ReqwestError: error sending request for url (https://192.168.64.3:8443/apis/babylontech.co.uk/v1/namespaces/apps/shipcatmanifests/dummy-s
ervice2/status?): error trying to connect: invalid dnsname: ReqwestError(reqwest::Error { kind: Request, url: "https://192.168.64.3:8443/apis/babylontech.co.uk/v1/namespaces/apps/shipcatmanifests/dummy-service2/status?", source: hyper::Error(Connect, Custom { kind: Other, error: "invalid dnsname" }) })
```

This is due to missing support for it upstream: https://github.com/ctz/hyper-rustls/issues/56

## In-cluster ~~broken~~ fixed
Cannot seem to get this to work within-cluster yet because we always get a straight IP to connect to.
At least in EKS I seem to just get a straight IP to connect to the api server. So this is pretty ununsable for me atm.

**Update**: in-cluster fixed in #597

## local minikube / k3d workaround
Add an ip entry (here `192.168.64.3 localhost`) to `/etc/hosts`, then and search replace the ip with `localhost` in `~/.kube/config`

**Comment by clux** [comment]
This is not really "fixed" but worked around in `master`. We now connect to the apiserver using cluster dns via solution suggested in #587 .

**Comment by mattsawyer77** [comment]
@clux there doesn't appear to be any workaround for accessing a cluster with a numeric hostname from the outside with rustls. Though I'm not sure kube-rs has any responsibility here, at least until the following issues are resolved?

* https://github.com/rustls/hyper-rustls/issues/56
* https://github.com/ctz/rustls/issues/184
* https://github.com/briansmith/webpki/issues/54

That said, should this remain open for tracking?

**Comment by clux** [comment]
Outside the cluster you can workaround it by tweaking `/etc/hosts` and making a non-ip alias. Not sure that will work for all cases though.

I think you're right and that even though it's by default worked around in-cluster, it still requires work outside.

**Comment by mattsawyer77** [comment]
@clux thanks for the quick response. Unfortunately, I don’t think I can override in my case, because the certs won’t match the hostname.

**Comment by flavio** [comment]
JFYI: the rustls folks have forked the `webpki` crate and their fork has a fix for this issue. You can see the issue about IP support here: https://github.com/rustls/webpki/issues/4

Currently there's no official release of this forked crate, plus a bunch of projects will need to switch to [`rustls-webpki`](https://crates.io/crates/rustls-webpki) instead of [`webpki`](https://crates.io/crates/webpki). Hence it will take time for this fix to be "consumable", but at least there's hope to get a proper fix :crossed_fingers:

**Comment by clux** [comment]
This does indeed appear fixed in `main` now. @flavio's efforts :heart: have trickled down through `rustls`, and our previously failing rustls ci checks are back #1183. So will close this issue in the hope that we can finally leave this multi-year saga behind :older_adult: 

Expect this change included in `kube@0.81.0`.
